src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamTests.cs

```diff
@@ -434,6 +435,138 @@ public async Task Read_StreamAborted_Throws()
+                    Assert.Equal(expectedErrorCode, ex.ErrorCode);
+
+                    await stream.ShutdownCompleted();
+                }
+            );
+        }
+
+        [Fact]
+        public async Task WritePreCanceled_Throws()
+        {
+            long expectedErrorCode = 1234;
+
+            await RunClientServer(
+                clientFunction: async connection =>
+                {
+                    await using QuicStream stream = connection.OpenUnidirectionalStream();
+
+                    CancellationTokenSource cts = new CancellationTokenSource();
+                    cts.Cancel();
+
+                    await Assert.ThrowsAsync<OperationCanceledException>(() => stream.WriteAsync(new byte[1], cts.Token).AsTask());
+
+                    // next write would also throw
+                    await Assert.ThrowsAsync<OperationCanceledException>(() => stream.WriteAsync(new byte[1]).AsTask());
+
+                    // manual write abort is still required
+                    stream.AbortWrite(expectedErrorCode);
+
+                    await stream.ShutdownCompleted();
+                },
+                serverFunction: async connection =>
+                {
+                    await using QuicStream stream = await connection.AcceptStreamAsync();
+
+                    byte[] buffer = new byte[1024 * 1024];
+
+                    // TODO: it should always throw QuicStreamAbortedException, but sometimes it does not https://github.com/dotnet/runtime/issues/53530
+                    //QuicStreamAbortedException ex = await Assert.ThrowsAsync<QuicStreamAbortedException>(() => ReadAll(stream, buffer));
+                    try
+                    {
+                        await ReadAll(stream, buffer);
+                    }
+                    catch (QuicStreamAbortedException) { }
+
+                    await stream.ShutdownCompleted();
+                }
+            );
+        }
+
+        [Fact]
+        public async Task WriteCanceled_NextWriteThrows()
+        {
+            long expectedErrorCode = 1234;
+
+            await RunClientServer(
+                clientFunction: async connection =>
+                {
+                    await using QuicStream stream = connection.OpenUnidirectionalStream();
+
+                    CancellationTokenSource cts = new CancellationTokenSource(500);
+
+                    async Task WriteUntilCanceled()
+                    {
+                        var buffer = new byte[64 * 1024];
+                        while (true)
+                        {
+                            await stream.WriteAsync(buffer, cancellationToken: cts.Token);
+                        }
+                    }
+
+                    // a write would eventually be canceled
+                    await Assert.ThrowsAsync<OperationCanceledException>(() => WriteUntilCanceled());
```

Comment:
Should we call `WriteUntilCanceled().WaitAsync(DefaultTestTimeout)` or something similar to ensure that if this does fail we get a proper exception instead of a hanging test? We'd have to be careful the ThrowsAsync doesn't swallow the test timeout.

Reply:
I will try that, or making a loop big enough but finite, as @ManickaP suggested, and see what will work best